sshd-core/src/main/java/org/apache/sshd/server/session/ServerUserAuthService.java

```diff
@@ -154,6 +154,26 @@ public synchronized void process(int cmd, Buffer buffer) throws Exception {
         ServerSession session = getServerSession();
         boolean debugEnabled = log.isDebugEnabled();
         if (cmd == SshConstants.SSH_MSG_USERAUTH_REQUEST) {
+            /*
+             * According to RFC4252 section 5.1:
+             *
+             *
+             *      When SSH_MSG_USERAUTH_SUCCESS has been sent, any
+             *      further authentication requests received after that
+             *      SHOULD be silently ignored.
+             */
+            if (session.isAuthenticated()) {
+                String username = buffer.getString();
+                String service = buffer.getString();
+                String method = buffer.getString();
+
+                if (debugEnabled) {
```

Comment:
Why this guard is necessary? I believe the methods themselves will have the level check.

Reply:
1. Yes, but:

```java
void debug(String format, Object ... args);
```
Which means that an intermediate array is automatically created, pushed to the stack and then eventually discarded - i.e., un-necessary object creating on the heap.

2. The same applies for
```java
log.debug("This " + session + " is " + getState() + " since " + getStartDate());
```
It creates a `StringBuilder`, appends to it various values (converting them to string(s)), converts it `toString()` and then does nothing with it - result: 2 (or more) objects created for nothing

3. Turns out that `isDebugEnabled()` is "cheaper" then calling `log.debug` and letting it figure out that there is nothing to do - even if we ignored the other 2 reasons.